### DISCUSSION TITLE
Slope and Linear Regression

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#241] [Indicators]
Community discussion about [Slope and Linear Regression](https://dotnet.stockindicators.dev/indicators/Slope/#content)

**Comment by DaveSkender** [comment]
From @ray2k:
> Any chance of getting a standard dev channel indicator? It's basically a best-fit linear regression and (typically 2, sometimes 1) standard deviations up/down from the line where (in the cast of 2 deviations) something like 95% of price action tends to take place. I think this can be pieced together now between the 'Slope and Linear Regression' and 'Standard Deviation' stuff but seems less than ideal.

**Reply by DaveSkender** [reply]
Have you seen this somewhere I can reference? We usually don't do custom indicators, but I think this is doable as an extension of Slope, much like we do for `GetSma` and `GetSmaExtended`.

**Reply by DaveSkender** [reply]
Were you thinking something like this with standard deviation for the reference window:
![image](https://user-images.githubusercontent.com/8432125/107866598-9e6f7900-6e40-11eb-8010-4247cde27c41.png)

Or like this, using the rolling standard deviation:
![image](https://user-images.githubusercontent.com/8432125/107866622-c65edc80-6e40-11eb-85bd-9d0fc0676dcf.png)

**Reply by ray2k** [reply]
I'm most familiar with the StandardDevChannel as it exists in thinkorswim: 
https://tlc.thinkorswim.com/center/reference/Tech-Indicators/studies-library/R-S/StandardDevChannel
also:
https://www.incrediblecharts.com/indicators/standard_deviation_channels.php
https://www.interactivebrokers.com/en/software/tws/usersguidebook/technicalanalytics/standardDeviation.htm

As far as your question, more like the first one. In thinkorswim it defaults to basing the trendline over the entire chart which is how I find it useful (then can switch chart timeperiod for different views/etc). Don't know enough if that is the standard/default or not.

**Reply by DaveSkender** [reply]
Ok, I see.  Since this is used in reputable broker sites, I think this would be a good indicator to add.  Please go ahead and create an Issue and I'll put it in the queue.

**Reply by DaveSkender** [reply]
I think what we can do is make the `lookbackPeriod` optional; that way, we can do "entire chart" by default.  See #361 for development notes and status.

UPDATE: This indicator has been implemented as [Standard Deviation Channels](https://daveskender.github.io/Stock.Indicators/indicators/StdDevChannels/#content)

**Comment by Snappee** [comment]
There are situations where the function does not compute the LINE value in the result but returns everything else.   I have a special use case where I first compute a momentum value, then pass it through linear regression to calculate the line, slope, etc..  I have not had time to look into your core code to figure out why a NULL exists, so currently I check for a NULL value in the line and do the standard "line = M*X + B" calc.

**Reply by DaveSkender** [reply]
It only computes the line over the last `N` periods for `quotes.GetSlope(N)`, so the `Line` value will be `null` for all periods in your results before the last `N` bars.

![45EC5510-A99C-4627-8092-41D12B783E9D](https://user-images.githubusercontent.com/8432125/204965857-4db1a136-3135-4b76-80b6-fd004bc15829.jpeg)

**Reply by DaveSkender** [reply]
Add the `Linear Regression` option here to test it out: https://charts.stockindicators.dev/
